content/browser/attribution_reporting/attribution_storage_sql.cc

```diff
@@ -2999,6 +3388,272 @@ bool AttributionStorageSql::StoreAttributionReport(AttributionReport& report) {
+      return AggregatableResult::kInsufficientBudget;
+    }
+  }
+
+  // Insert epochs not already stored
+  static constexpr char kInsertEpochOrigin[] =
+    "INSERT INTO per_origin_filters"
+    "(epoch,origin,initial_budget,consumed_budget)"
+    "VALUES(?,?,?,?)";
+  
+  for (auto& epoch : epochs_to_insert) {
+    sql::Statement stmt(db_.GetCachedStatement(SQL_FROM_HERE, kInsertEpochOrigin));
+    stmt.BindInt64(0, epoch);
+    stmt.BindString(1, net::SchemefulSite(querying_origin).Serialize());
+    stmt.BindDouble(2, kInitialBudget);
+    stmt.BindDouble(3, required_budget);
+    if (!stmt.Run()) {
+      return AggregatableResult::kInternalError;
+    }  
+  }
+  
+  // Update epochs already stored
+  static constexpr char kAdjustBudgetConsumed[] =
+    "UPDATE per_origin_filters "
+    "SET consumed_budget=consumed_budget+?"
+    "WHERE epoch=? and origin=?";
+
+  for (auto& epoch : epochs_to_update) {
+    sql::Statement stmt(db_.GetCachedStatement(SQL_FROM_HERE, kAdjustBudgetConsumed));
+    stmt.BindDouble(0, required_budget);
+    stmt.BindInt64(1, epoch);
+    stmt.BindString(2, net::SchemefulSite(querying_origin).Serialize());
+    if (!stmt.Run()) {
+      return AggregatableResult::kInternalError;
+    }
+  }
+
+  // TODO(kelly): optimize, don't do insertions/updates in a loop?
+  
+  // Finishing atomic operation
+  if (!transaction.Commit()) {
+    return AggregatableResult::kInternalError;
+  }
+
+  return AggregatableResult::kSuccess;
+}
+
+
+AggregatableResult
+AttributionStorageSql::MaybeStoreAggregatableAttributionReportDataM2M(
+    const AttributionInfo& attribution_info,
+    std::vector<Partition>& partitions,
+    std::optional<AttributionReport>& report,
+    const AttributionTrigger& trigger,
+    StoredSource& source /*for retrocompatibility*/) {
+
+  // Computes the budget-required per requested epoch for the querying origin using 
+  // different optimization methods. Tries to consume budget from every requested epoch from the
+  // querying origin. Partitions whose epochs don't have remaining budget will have their reports NULLed.
+  // Sums the reports across partitions to create one attribution report.
+
+  const SuitableOrigin& querying_origin = trigger.destination_origin();
+
+  const attribution_reporting::TriggerRegistration& trigger_registration =
+      trigger.registration();
+
+  const attribution_reporting::AggregatableValues& aggregatable_cap_values = 
+      trigger_registration.aggregatable_cap_values;
+
+  const double global_epsilon = trigger_registration.global_epsilon;
+
+  // Compute global sensitivity
+  double global_sensitivity = 0;
+  if (std::strcmp(kSensitivityMetric, "L1") == 0) {
+    for (auto& pair : aggregatable_cap_values.values()) {
+      global_sensitivity += pair.second;
+    }
+  } else if (std::strcmp(kSensitivityMetric, "L2") == 0) {
+    // TODO(kelly)
+    // sum up histograms, raise each bucket value to power of 2 sum them up and sqrt
```

Comment:
Yes but let's wait for that one to be sure that we don't mess up with a non-linearity due to the square/sqrt